src/nanoemoji/parts.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import dataclasses
+from functools import lru_cache
+import json
+from nanoemoji.config import FontConfig
+from pathlib import Path
+from picosvg.svg import SVG
+from picosvg.svg_reuse import normalize
+from picosvg.svg_types import SVGPath, SVGShape
+from typing import Iterable, List, MutableMapping, Set, Tuple, Union
+
+
+PathSource = Union[SVGShape, Iterable[SVGShape], "ReuseableParts", Path]
+
+
+@lru_cache(maxsize=1)
+def _default_tolerence() -> float:
+    return FontConfig().reuse_tolerance
+
+
+def _is_iterable_of(thing, desired_type) -> bool:
+    try:
+        it = iter(thing)
+    except TypeError:
+        return False
+
+    try:
+        val = next(it)
+        return isinstance(val, desired_type)
+    except StopIteration:
+        return True
+
+
+@dataclasses.dataclass(frozen=True)
+class Shape:
+    d: str  # an SVG style path, e.g. the d attribute of <svg:path/>
+
+
+# A set of shapes that normalize to the same path
+# Only the d attribute of the SVGPath is populated
+@dataclasses.dataclass
+class ShapeSet:
+    normalized: Shape
+    shapes: Set[Shape] = dataclasses.field(default_factory=set)
+
+
+@dataclasses.dataclass
+class ReuseableParts:
+    version: Tuple[int, int, int] = (1, 0, 0)
+    reuse_tolerance: float = dataclasses.field(default_factory=_default_tolerence)
+    shape_sets: MutableMapping[Shape, ShapeSet] = dataclasses.field(
+        default_factory=dict
+    )
+
+    def _add_norm_path(self, norm: Shape, shape: Shape):
+        if norm not in self.shape_sets:
+            self.shape_sets[norm] = ShapeSet(norm)
+        self.shape_sets[norm].shapes.add(shape)
+
+    def _add(self, shape: Shape):
+        norm = shape
+        if self.reuse_tolerance != -1:
+            norm = Shape(normalize(SVGPath(d=shape.d), self.reuse_tolerance).d)
+        self._add_norm_path(norm, shape)
+
+    def add(self, source: PathSource):
+        if isinstance(source, Path):
+            source = ReuseableParts.load(source)
+
+        if isinstance(source, ReuseableParts):
```

Comment:
I wonder if it would it be too clever to also override the `__or__`, `__ior__` magic methods to be able to do `parts = p1 | p2` or `parts |= other_parts`

then i'd leave the `add` method for single SVGShapes

Reply:
I waffled on whether that would be pythonic and ultimately felt it was a little on the obtuse side. WDYT?

(I might clean up other feedback and merge, we can always make this change later if we like it)